File: assign6/apriori.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <set>
#include <map>
#include <algorithm>
#include <string>

using namespace std;
void writeCSV(const string &filename, const vector<vector<string>> &rows)
{
    ofstream file(filename);
    for (auto &row : rows)
    {
        for (size_t i = 0; i < row.size(); i++)
        {
            file << row[i];
            if (i != row.size() - 1)
                file << ",";
        }
        file << "\n";
    }
    file.close();
}
vector<set<string>> generateCandidates(const vector<set<string>> &prevFreq, int k)
{
    vector<set<string>> candidates;
    for (size_t i = 0; i < prevFreq.size(); i++)
    {
        for (size_t j = i + 1; j < prevFreq.size(); j++)
        {
            set<string> c;
            set_union(prevFreq[i].begin(), prevFreq[i].end(),
                      prevFreq[j].begin(), prevFreq[j].end(),
                      inserter(c, c.begin()));
            if ((int)c.size() == k)
                candidates.push_back(c);
        }
    }
    sort(candidates.begin(), candidates.end());
    candidates.erase(unique(candidates.begin(), candidates.end()), candidates.end());
    return candidates;
}
map<set<string>, int> countSupport(const vector<set<string>> &transactions,
                                   const vector<set<string>> &candidates)
{
    map<set<string>, int> supportCount;
    for (auto &c : candidates)
    {
        for (auto &t : transactions)
        {
            if (includes(t.begin(), t.end(), c.begin(), c.end()))
            {
                supportCount[c]++;
            }
        }
    }
    return supportCount;
}
void apriori(const vector<set<string>> &transactions, double minSupport, double minConfidence)
{
    int totalTransactions = transactions.size();
    vector<set<string>> oneItemsets;
    set<string> allItems;
    for (auto &t : transactions)
    {
        for (auto &item : t)
            allItems.insert(item);
    }
    for (auto &item : allItems)
        oneItemsets.push_back({item});

    vector<vector<set<string>>> allFrequentSets;
    int k = 1;

    vector<vector<string>> freqCSV = {{"Itemset", "Support"}};
    vector<vector<string>> rulesCSV = {{"Antecedent", "Consequent", "Support", "Confidence", "Lift"}};

    while (true)
    {
        map<set<string>, int> supportCount = countSupport(transactions, oneItemsets);

        vector<set<string>> freqItemsets;
        for (auto &p : supportCount)
        {
            double support = (double)p.second / totalTransactions;

            if (support >= minSupport)
            {
                freqItemsets.push_back(p.first);
                string items = "";
                for (auto &i : p.first)
                    items += i + " ";
                freqCSV.push_back({items, to_string(support)});
            }
        }
        if (freqItemsets.empty())
            break;

        allFrequentSets.push_back(freqItemsets);

        oneItemsets = generateCandidates(freqItemsets, ++k);
    }
    cout << "\nFrequent Itemsets (minSup=" << minSupport << ") \n";
    for (size_t i = 0; i < allFrequentSets.size(); i++)
    {
        for (auto &s : allFrequentSets[i])
        {
            for (auto &item : s)
                cout << item << " ";
            cout << endl;
        }
    }
    cout << "\nAssociation Rules (minSup=" << minSupport << ", minConf=" << minConfidence << ")\n";
    for (auto &level : allFrequentSets)
    {
        for (auto &itemset : level)
        {
            if (itemset.size() < 2)
                continue;

            vector<string> items(itemset.begin(), itemset.end());
            int n = items.size();
            for (int mask = 1; mask < (1 << n) - 1; mask++)
            {
                set<string> antecedent, consequent;
                for (int i = 0; i < n; i++)
                {
                    if (mask & (1 << i))
                        antecedent.insert(items[i]);
                    else
                        consequent.insert(items[i]);
                }

                map<set<string>, int> supCount = countSupport(transactions, {itemset, antecedent});
                double supportItemset = (double)supCount[itemset] / totalTransactions;
                double supportAntecedent = (double)supCount[antecedent] / totalTransactions;

                if (supportAntecedent == 0)
                    continue;
                double confidence = supportItemset / supportAntecedent;

                double supportConsequent = (double)countSupport(transactions, {consequent})[consequent] / totalTransactions;
                if (supportConsequent == 0)
                    continue;
                double lift = confidence / supportConsequent;
                if (supportItemset >= minSupport && confidence >= minConfidence)
                {
                    cout << "{ ";
                    for (auto &a : antecedent)
                        cout << a << " ";
                    cout << "} => { ";
                    for (auto &c : consequent)
                        cout << c << " ";
                    cout << "} (support=" << supportItemset
                         << ", confidence=" << confidence
                         << ", lift=" << lift << ")\n";

                    string ant = "", con = "";
                    for (auto &a : antecedent)
                        ant += a + " ";
                    for (auto &c : consequent)
                        con += c + " ";
                    rulesCSV.push_back({ant, con, to_string(supportItemset),
                                        to_string(confidence), to_string(lift)});
                }
            }
        }
    }
    writeCSV("frequent_itemsets.csv", freqCSV);
    writeCSV("association_rules.csv", rulesCSV);

    cout << "\nCSV files generated: frequent_itemsets.csv, association_rules.csv\n";
}
vector<set<string>> readTransactions(const string &filename)
{
    vector<set<string>> transactions;
    ifstream file(filename);
    string line;
    while (getline(file, line))
    {
        set<string> transaction;
        stringstream ss(line);
        string item;
        while (getline(ss, item, ','))
        {
            if (!item.empty())
                transaction.insert(item);
        }
        if (!transaction.empty())
            transactions.push_back(transaction);
    }
    return transactions;
}

int main()
{
    string inputFile = "transactions.csv"; 
    double minSupport = 0.4;    
    double minConfidence = 0.7; 

    vector<set<string>> transactions = readTransactions(inputFile);
    if (transactions.empty())
    {
        cout << "No transactions found in " << inputFile << endl;
        return 0;
    }

    apriori(transactions, minSupport, minConfidence);
    return 0;
}

File: assign6/Apriori.java
import java.io.*;
import java.util.*;

public class Apriori {

    // Read CSV file and return transactions
    public static List<Set<String>> readCSV(String filename) throws IOException {
        List<Set<String>> transactions = new ArrayList<>();
        BufferedReader br = new BufferedReader(new FileReader(filename));
        String line;
        while ((line = br.readLine()) != null) {
            Set<String> transaction = new HashSet<>(Arrays.asList(line.split(",")));
            transactions.add(transaction);
        }
        br.close();
        return transactions;
    }

    // Generate candidate 1-itemsets
    public static Set<Set<String>> genCan(List<Set<String>> transactions) {
        Set<Set<String>> candidate1 = new HashSet<>();
        for (Set<String> transaction : transactions) {
            for (String item : transaction) {
                candidate1.add(new HashSet<>(Collections.singletonList(item)));
            }
        }
        return candidate1;
    }

    // Count support for candidate itemsets
    public static Map<Set<String>, Integer> countSupport(List<Set<String>> transactions, Set<Set<String>> candidates) {
        Map<Set<String>, Integer> supportCounts = new HashMap<>();
        for (Set<String> transaction : transactions) {
            for (Set<String> candidate : candidates) {
                if (transaction.containsAll(candidate)) {
                    supportCounts.put(candidate, supportCounts.getOrDefault(candidate, 0) + 1);
                }
            }
        }
        return supportCounts;
    }

    // Prune based on minSupport
    public static Set<Set<String>> pruneItemsets(Map<Set<String>, Integer> supportCounts, double minSupport, int totalTransactions) {
        Set<Set<String>> frequentItemsets = new HashSet<>();
        for (Map.Entry<Set<String>, Integer> entry : supportCounts.entrySet()) {
            double supportPercentage = (double) entry.getValue() / totalTransactions * 100;
            if (supportPercentage >= minSupport) {
                frequentItemsets.add(entry.getKey());
            }
        }
        return frequentItemsets;
    }

    // Generate candidate k-itemsets
    public static Set<Set<String>> generateCandidates(Set<Set<String>> frequentItemsets, int k) {
        Set<Set<String>> candidates = new HashSet<>();
        List<Set<String>> itemsets = new ArrayList<>(frequentItemsets);

        for (int i = 0; i < itemsets.size(); i++) {
            for (int j = i + 1; j < itemsets.size(); j++) {
                Set<String> merged = new HashSet<>(itemsets.get(i));
                merged.addAll(itemsets.get(j));
                if (merged.size() == k) {
                    candidates.add(merged);
                }
            }
        }
        return candidates;
    }

    // Print frequent itemsets
    public static void printFrequentItemsets(Set<Set<String>> frequentItemsets, Map<Set<String>, Integer> supportCounts, int totalTransactions) {
        for (Set<String> itemset : frequentItemsets) {
            double supportPercentage = (double) supportCounts.get(itemset) / totalTransactions * 100;
            System.out.print("{ ");
            for (String item : itemset) {
                System.out.print(item + " ");
            }
            System.out.println("} - Support: " + supportPercentage + "%");
        }
    }

    // Generate association rules
    public static void generateAssociationRules(Set<Set<String>> allFrequentItemsets, Map<Set<String>, Integer> supportCounts, int totalTransactions, double minConfidence) {
        for (Set<String> itemset : allFrequentItemsets) {
            if (itemset.size() < 2) continue;

            List<String> items = new ArrayList<>(itemset);
            int n = items.size();
            int supportItemset = supportCounts.get(itemset);

            for (int mask = 1; mask < (1 << n) - 1; mask++) {
                Set<String> antecedent = new HashSet<>();
                Set<String> consequent = new HashSet<>();
                for (int i = 0; i < n; i++) {
                    if ((mask & (1 << i)) != 0)
                        antecedent.add(items.get(i));
                    else
                        consequent.add(items.get(i));
                }

                if (!supportCounts.containsKey(antecedent)) continue;
                int supportAntecedent = supportCounts.get(antecedent);
                double confidence = (double) supportItemset / supportAntecedent * 100;

                if (confidence >= minConfidence) {
                    double supportPercentage = (double) supportItemset / totalTransactions * 100;
                    System.out.print("{ ");
                    for (String a : antecedent) System.out.print(a + " ");
                    System.out.print("} => { ");
                    for (String c : consequent) System.out.print(c + " ");
                    System.out.println("} (Support: " + supportPercentage + "%, Confidence: " + confidence + "%)");
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter CSV file name: ");
        String filename = sc.nextLine();
        System.out.print("Enter minimum support (%): ");
        double minSupport = sc.nextDouble();
        System.out.print("Enter minimum confidence (%): ");
        double minConfidence = sc.nextDouble();
        sc.close();

        List<Set<String>> transactions = readCSV(filename);
        int totalTransactions = transactions.size();

        Set<Set<String>> candidate1 = genCan(transactions);
        Map<Set<String>, Integer> supportCounts = countSupport(transactions, candidate1);
        Set<Set<String>> allFrequentItemsets = new HashSet<>();
        Set<Set<String>> frequentItemsets = pruneItemsets(supportCounts, minSupport, totalTransactions);
        allFrequentItemsets.addAll(frequentItemsets);

        System.out.println("Frequent Itemsets (size 1):");
        printFrequentItemsets(frequentItemsets, supportCounts, totalTransactions);

        int k = 2;
        while (!frequentItemsets.isEmpty()) {
            Set<Set<String>> candidateK = generateCandidates(frequentItemsets, k);
            if (candidateK.isEmpty()) break;
            Map<Set<String>, Integer> supportK = countSupport(transactions, candidateK);
            frequentItemsets = pruneItemsets(supportK, minSupport, totalTransactions);

            if (!frequentItemsets.isEmpty()) {
                System.out.println("\nFrequent Itemsets (size " + k + "):");
                printFrequentItemsets(frequentItemsets, supportK, totalTransactions);

                supportCounts.putAll(supportK);
                allFrequentItemsets.addAll(frequentItemsets);
            }
            k++;
        }

        System.out.println("\nAssociation Rules:");
        generateAssociationRules(allFrequentItemsets, supportCounts, totalTransactions, minConfidence);
    }
}
